src/Init/Lean/Delaborator.lean

```diff
@@ -0,0 +1,371 @@
+elaborators for `app.c` are tried first; this is also done for `Expr.const`s (\"nullary applications\")
+to reduce special casing. If the term is an `Expr.mdata` with a single key `k`, `mdata.k`
+is tried first.",
+  valueTypeName := `Lean.Delaborator.Delab
+} `Lean.Delaborator.delabAttribute
+@[init mkDelabAttribute] constant delabAttribute : KeyedDeclsAttribute Delab := arbitrary _
+
+def getExpr : DelabM Expr := do
+ctx ← read;
+pure ctx.expr
+
+def getExprKind : DelabM Name := do
+e ← getExpr;
+pure $ match e with
+| Expr.bvar _ _        => `bvar
+| Expr.fvar _ _        => `fvar
+| Expr.mvar _ _        => `mvar
+| Expr.sort _ _        => `sort
+| Expr.const c _ _     =>
+  -- we identify constants as "nullary applications" to reduce special casing
+  `app ++ c
+| Expr.app fn _ _      => match fn.getAppFn with
+  | Expr.const c _ _ => `app ++ c
+  | _                => `app
+| Expr.lam _ _ _ _     => `lam
+| Expr.forallE _ _ _ _ => `forallE
+| Expr.letE _ _ _ _ _  => `letE
+| Expr.lit _ _         => `lit
+| Expr.mdata m _ _     => match m.entries with
+  | [(key, _)] => `mdata ++ key
+  | _   => `mdata
+| Expr.proj _ _ _ _    => `proj
+| Expr.localE _ _ _ _  => `localE
+
+/-- Evaluate option accessor, using subterm-specific options if set. Default to `true` if `pp.all` is set. -/
+def getPPOption (opt : Options → Bool) : DelabM Bool := do
+ctx ← read;
+let opt := fun opts => opt opts || getPPAll opts;
+let val := opt ctx.defaultOptions;
+match ctx.optionsPerPos.find? ctx.pos with
+| some opts => pure $ opt opts
+| none      => pure val
+
+def whenPPOption (opt : Options → Bool) (d : Delab) : Delab := do
+b ← getPPOption opt;
+if b then d else failure
+
+def whenNotPPOption (opt : Options → Bool) (d : Delab) : Delab := do
+b ← getPPOption opt;
+if b then failure else d
+
+/-- Descend into subterm `child`, the `childIdex`-th out of `numChildren` subterms, and update position. -/
+def descend {α} (child : Expr) (childIdx numChildren : Nat) (d : DelabM α) : DelabM α :=
+adaptReader (fun (cfg : Context) => { cfg with expr := child, pos := (cfg.pos + childIdx) * numChildren }) d
+
+def withAppFn {α} (d : DelabM α) : DelabM α := do
+Expr.app fn _ _ ← getExpr | unreachable!;
+descend fn 0 2 d
+
+def withAppArg {α} (d : DelabM α) : DelabM α := do
+Expr.app _ arg _ ← getExpr | unreachable!;
+descend arg 1 2 d
+
+partial def withAppFnArgs {α} : DelabM α → (α → DelabM α) → DelabM α
+| fnD, argD => do
+  Expr.app fn arg _ ← getExpr | fnD;
+  a ← withAppFn (withAppFnArgs fnD argD);
+  withAppArg (argD a)
+
+def withBindingDomain {α} (d : DelabM α) : DelabM α := do
+e ← getExpr;
+descend e.bindingDomain! 0 2 d
+
+def withBindingBody {α} (n : Name) (d : DelabM α) : DelabM α := do
+e ← getExpr;
+fun ctx => withLocalDecl n e.bindingDomain! e.binderInfo $ fun fvar =>
+  let b := e.bindingBody!.instantiate1 fvar;
+  descend b 1 2 d ctx
+
+def infoForPos (pos : Nat) : SourceInfo :=
+{ leading := " ".toSubstring, pos := pos, trailing := " ".toSubstring }
```

Comment:
Bad hack: we cannot set the position without also setting the whitespace. I think we'll need to change that.

Reply:
What about `updatePos (stx : Syntax) (pos : Nat) : Syntax`? 
Then, we can write `stx.setInfo (infoForPos pos)` as `stx.updatePos pos`.